providers/gpgpu/bin/gpu_passthrough.py

```diff
@@ -0,0 +1,576 @@
+            filename = os.path.join("/tmp", targetfile)
+            if not os.path.isfile(filename):
+                self.download_image(self.image_url, filename)
+            self._image = filename
+
+        return self._image
+
+    def run(
+        self,
+        cmd: str,
+        log_stderr: bool = True,
+        on_guest: bool = False,
+        check: bool = False,
+    ) -> subprocess.CompletedProcess:
+        """Runs a command on the host or instance."""
+        if on_guest:
+            cmd = "lxc exec {} -- {}".format(self.name, cmd)
+        proc = subprocess.run(
+            shlex.split(cmd),
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            stdin=subprocess.DEVNULL,
+            universal_newlines=True,
+            check=False,
+        )
+
+        logging.debug("Command: %s", cmd)
+        if proc.stdout:
+            logging.debug(" STDOUT: %s", proc.stdout)
+        if log_stderr and proc.stderr:
+            logging.debug(" STDERR: %s", proc.stderr)
+
+        if check:
+            proc.check_returncode()
+
+        return proc
+
+    def download_image(self, url, filename):
+        """Downloads LXD files for same release as host machine."""
+        logging.debug("Attempting download of %s from %s", filename, url)
+        try:
+            urllib.request.urlretrieve(url, filename)
+        except (
+            IOError,
+            OSError,
+            urllib.error.HTTPError,
+            urllib.error.URLError,
+        ) as e:
+            logging.exception(e)
+        except ValueError as verr:
+            logging.exception(verr)
+
+        if not os.path.isfile(filename):
+            raise FileNotFoundError(filename)
+
+    def insert_images(self):
+        """Insert LXD template and image images."""
+        if self.template and self.image:
+            logging.debug("Importing images into LXD")
+            self.run(
+                "lxc image import {} rootfs {} --alias {}".format(
+                    self.template, self.image, self.image_alias.hex
+                ),
+                check=True,
+            )
+        else:
+            logging.debug("No local images, attempting import from remote")
+            retries = 2
+            for _ in range(retries):
+                proc = self.run(
+                    "lxc image copy {}{} local: --alias {}".format(
+                        self.remote, RELEASE, self.image_alias.hex
+                    )
+                )
+                if proc.returncode == 0:
+                    return
+                logging.error("Error encountered while importing images")
+                logging.error("Attempting up to %d times", retries)
+            raise RuntimeError("Images could not be inserted")
+
+    def init_lxd(self):
+        """Initializes LXD."""
+        if self.run("lxd waitready --timeout=5").returncode != 0:
+            self.run("lxd init --auto", check=True)
```

Comment:
We switched to exceptions, so lets change this as well:
```suggestion
        try:
            self.run("lxd waitready --timeout=5")
        except CalledProcessError:
            self.run("lxd init --auto")
```